### PR TITLE
[FLINK-30389] Add retry to read hints

### PR DESCRIPTION
**Pull request by WencongLiu**
## What is the purpose of the change

Add retry to read hints in SnapshotManager

## Brief change log

  - *Introduce the retry logic in SnapshotManager*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no )
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (JavaDocs)

### CODE REVIEW

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/utils/SnapshotManager.java [hunk 1]
```diff
@@ -214,13 +217,22 @@ public Optional<Snapshot> latestSnapshotOfUser(String user) {
     public Long readHint(String fileName) {
         Path snapshotDir = snapshotDirectory();
         Path path = new Path(snapshotDir, fileName);
+        int retryNumber = 0;
         try {
-            if (path.getFileSystem().exists(path)) {
-                return Long.parseLong(FileUtils.readFileUtf8(path));
+            while (retryNumber++ < READ_HINT_RETRY_NUM) {
+                if (path.getFileSystem().exists(path)) {
+                    return Long.parseLong(FileUtils.readFileUtf8(path));
```
Comment (JingsongLi) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/file/utils/SnapshotManager.java:224]: We need to catch exception and ignore exception here.

Comment (WencongLiu) [reply flink-table-store-core/src/main/java/org/apache/flink/table/store/file/utils/SnapshotManager.java:224]: Done.

Comment (JingsongLi) [reply flink-table-store-core/src/main/java/org/apache/flink/table/store/file/utils/SnapshotManager.java:224]:
```
int retryNumber = 0;
while (retryNumber++ < READ_HINT_RETRY_NUM) {
        try {
             return Long.parseLong(FileUtils.readFileUtf8(path));
        } catch (Exception ignored) {
        }
       TimeUnit.MILLISECONDS.sleep(READ_HINT_RETRY_INTERVAL);
}
```

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/utils/SnapshotManager.java [hunk 2]
```diff
@@ -44,6 +45,8 @@ public class SnapshotManager {
     private static final String SNAPSHOT_PREFIX = "snapshot-";
     public static final String EARLIEST = "EARLIEST";
     public static final String LATEST = "LATEST";
+    private static final int READ_HINT_RETRY_NUM = 10;
+    private static final int READ_HINT_RETRY_INTERVAL = 200;
```
Comment (JingsongLi) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/file/utils/SnapshotManager.java:49]: Maybe the better choice is:
- `READ_HINT_RETRY_NUM` is 3.
- `READ_HINT_RETRY_INTERVAL` is 1ms.

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/utils/SnapshotManager.java (hunk 1 above)
Comment (WencongLiu) [reply flink-table-store-core/src/main/java/org/apache/flink/table/store/file/utils/SnapshotManager.java:224]: OK. I've added a fix-up commit. Currently only exception of "Long.parseLong(FileUtils.readFileUtf8(path))" is catched, and others are thrown to invoker.

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/utils/SnapshotManager.java (hunk 2 above)
Comment (WencongLiu) [reply flink-table-store-core/src/main/java/org/apache/flink/table/store/file/utils/SnapshotManager.java]: Done.

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/utils/SnapshotManager.java (hunk 1 above)
Comment (JingsongLi) [reply flink-table-store-core/src/main/java/org/apache/flink/table/store/file/utils/SnapshotManager.java:224]: Maybe we can just remove `if (path.getFileSystem().exists(path)) {`, we already have catch the exception for `FileUtils.readFileUtf8(path)`.

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/utils/SnapshotManager.java [hunk 3]
```diff
@@ -211,16 +214,18 @@ public Optional<Snapshot> latestSnapshotOfUser(String user) {
...
-                return Long.parseLong(FileUtils.readFileUtf8(path));
+                try {
+                    return Long.parseLong(FileUtils.readFileUtf8(path));
+                } catch (IOException ignored) {
+                }
             }
-        } catch (Exception e) {
-            LOG.info(
-                    "Failed to read hint file " + fileName + ". Falling back to listing files.", e);
+            TimeUnit.MILLISECONDS.sleep(READ_HINT_RETRY_INTERVAL);
```
Comment (JingsongLi) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/file/utils/SnapshotManager.java:228]: Catch and ignore the `InterruptedException`

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/utils/SnapshotManager.java (hunk 1 above)
Comment (WencongLiu) [reply flink-table-store-core/src/main/java/org/apache/flink/table/store/file/utils/SnapshotManager.java]: Done.

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/utils/SnapshotManager.java (hunk 3 above)
Comment (WencongLiu) [reply flink-table-store-core/src/main/java/org/apache/flink/table/store/file/utils/SnapshotManager.java]: Done.

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/utils/SnapshotManager.java [hunk 4]
```diff
@@ -214,13 +217,16 @@ public Optional<Snapshot> latestSnapshotOfUser(String user) {
     public Long readHint(String fileName) {
         Path snapshotDir = snapshotDirectory();
         Path path = new Path(snapshotDir, fileName);
-        try {
-            if (path.getFileSystem().exists(path)) {
+        int retryNumber = 0;
+        while (retryNumber++ < READ_HINT_RETRY_NUM) {
+            try {
                 return Long.parseLong(FileUtils.readFileUtf8(path));
+            } catch (IOException ignored) {
```
Comment (JingsongLi) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/file/utils/SnapshotManager.java:224]: `Exception` is more safe, after all, read hint is just an optimization

**Review by JingsongLi — COMMENTED** [review]
Hi @WencongLiu  thanks for the update, just one comment.

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/utils/SnapshotManager.java (hunk 4 above)
Comment (WencongLiu) [reply flink-table-store-core/src/main/java/org/apache/flink/table/store/file/utils/SnapshotManager.java]: Done.

**Review by JingsongLi — APPROVED** [review]
Looks good to me! Thanks @WencongLiu

**Pull request merged**